cpp/src/arrow/compute/kernels/vector_cumulative_ops.cc

```diff
@@ -0,0 +1,183 @@
+    const auto& options = OptionsWrapper<OptionsType>::Get(ctx);
+    auto skip_nulls = options.skip_nulls;
+
+    auto out_type = batch.GetDescriptors()[0].type;
+    ARROW_ASSIGN_OR_RAISE(auto cast_value, Cast(Datum(options.start), out_type));
+    auto start = UnboxScalar<OutType>::Unbox(*cast_value.scalar());
+
+    int64_t base_output_offset = 0;
+    bool encounted_null = false;
+    ArrayData* out_arr = out->mutable_array();
+
+    switch (batch[0].kind()) {
+      case Datum::ARRAY: {
+        auto st = Call(ctx, base_output_offset, *batch[0].array(), out_arr, &start,
+                       skip_nulls, &encounted_null);
+        out_arr->SetNullCount(arrow::kUnknownNullCount);
+        return st;
+      }
+      case Datum::CHUNKED_ARRAY: {
+        const auto& input = batch[0].chunked_array();
+
+        for (const auto& chunk : input->chunks()) {
+          RETURN_NOT_OK(Call(ctx, base_output_offset, *chunk->data(), out_arr, &start,
+                             skip_nulls, &encounted_null));
+          base_output_offset += chunk->length();
+        }
+        out_arr->SetNullCount(arrow::kUnknownNullCount);
+        return Status::OK();
+      }
+      default:
+        return Status::NotImplemented(
+            "Unsupported input type for function 'cumulative_<operator>': ",
+            batch[0].ToString());
+    }
+  }
+
+  static Status Call(KernelContext* ctx, int64_t base_output_offset,
+                     const ArrayData& input, ArrayData* output, ArgValue* accumulator,
+                     bool skip_nulls, bool* encounted_null) {
+    Status st = Status::OK();
+    auto out_bitmap =
+        output->GetMutableValues<uint8_t>(0) + (base_output_offset / CHAR_BIT);
+    int64_t curr = base_output_offset % CHAR_BIT;
+    auto out_data = output->GetMutableValues<OutValue>(1) + base_output_offset;
+
+    auto null_func = [&]() {
+      *out_data++ = OutValue{};
+      *encounted_null = true;
+      arrow::bit_util::SetBitsTo(out_bitmap, curr, 1, false);
+      ++curr;
+    };
+
+    if (skip_nulls) {
+      VisitArrayValuesInline<ArgType>(
+          input,
+          [&](ArgValue v) {
+            *accumulator = Op::template Call<OutValue, ArgValue, ArgValue>(
+                ctx, v, *accumulator, &st);
+            *out_data++ = *accumulator;
+            ++curr;
+          },
+          null_func);
+    } else {
+      auto start_null_idx = curr;
+      VisitArrayValuesInline<ArgType>(
+          input,
+          [&](ArgValue v) {
+            if (*encounted_null) {
+              *out_data++ = OutValue{};
+            } else {
+              *accumulator = Op::template Call<OutValue, ArgValue, ArgValue>(
+                  ctx, v, *accumulator, &st);
+              *out_data++ = *accumulator;
+              ++start_null_idx;
+            }
+            ++curr;
+          },
+          null_func);
+
+      auto null_length = curr - start_null_idx;
```

Comment:
Isn't this effectively `input.length`? After all, we're writing out one element for each input.